Add alpha quality support for rds DBCluster

hiteshghia opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes
Add support for dbCluster in RDS using ACK code gen.
Steps followed till now -
<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #272

I have:

- [x] Read and followed Crossplane's [contribution process]. 
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
Manual testing is in progress.
<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

hiteshghia commented:
@muvaf opened this mr to replace https://github.com/crossplane/provider-aws/pull/489 and have  addressed all comments from there.
Going to work on manually testing it as the next step.

muvaf commented on apis/rds/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,64 @@
+import (
+	"context"
+
+	"github.com/crossplane/crossplane-runtime/pkg/reference"
+
+	iamv1beta1 "github.com/crossplane/provider-aws/apis/identity/v1beta1"
+
+	kmsv1alpha1 "github.com/crossplane/provider-aws/apis/kms/v1alpha1"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
```suggestion
	"github.com/pkg/errors"
	"sigs.k8s.io/controller-runtime/pkg/client"
	
	"github.com/crossplane/crossplane-runtime/pkg/reference"

	iamv1beta1 "github.com/crossplane/provider-aws/apis/identity/v1beta1"
	kmsv1alpha1 "github.com/crossplane/provider-aws/apis/kms/v1alpha1"
```

We go with this grouping starting from bottom `this repo -> repos in this org -> third part -> std`

hiteshghia-splunk replied:
ya for some reason lint was complaining, may be whitespace or  something.

hiteshghia replied:
ya for some reason lint was complaining, may be whitespace or something.

muvaf commented on apis/rds/v1alpha1/zz_db_cluster.go:
```diff
@@ -0,0 +1,475 @@
+	// default.
+	//
+	// Example: mySubnetgroup
+	DBSubnetGroupName *string `json:"dbSubnetGroupName,omitempty"`
```
This can be a reference to `DBSubnetGroup` in current `database/v1beta1` package.

muvaf commented on apis/rds/v1alpha1/zz_db_cluster.go:
```diff
@@ -0,0 +1,475 @@
+	// printable ASCII character except "/", """, or "@".
+	//
+	// Constraints: Must contain from 8 to 41 characters.
+	MasterUserPassword *string `json:"masterUserPassword,omitempty"`
```
We need to ignore this field and have a `Secret` reference to populate it. See the [example](https://github.com/crossplane/provider-aws/blob/087f587/apis/database/v1beta1/rdsinstance_types.go#L446) from `RDSInstance`.

muvaf commented on apis/rds/v1alpha1/zz_db_cluster.go:
```diff
@@ -0,0 +1,475 @@
+	// Tags to assign to the DB cluster.
+	Tags []*Tag `json:"tags,omitempty"`
+	// A list of EC2 VPC security groups to associate with this DB cluster.
+	VPCSecurityGroupIDs       []*string `json:"vpcSecurityGroupIDs,omitempty"`
```
This can be reference to `SecurityGroup` CRD.

muvaf commented on pkg/controller/rds/dbcluster/setup.go:
```diff
@@ -0,0 +1,75 @@
+	svcapitypes "github.com/crossplane/provider-aws/apis/rds/v1alpha1"
+)
+
+// SetupDbCluster adds a controller that reconciles DbCluster.
+func SetupDbCluster(mgr ctrl.Manager, l logging.Logger) error {
```
```suggestion
// SetupDBCluster adds a controller that reconciles DBCluster.
func SetupDBCluster(mgr ctrl.Manager, l logging.Logger) error {
```

muvaf commented on apis/rds/v1alpha1/generator-config.yaml:
```diff
@@ -0,0 +1,24 @@
+    - CreateDBClusterInput.DBClusterIdentifier
+    - ModifyDBClusterInput.DBClusterIdentifier
+    - DeleteDBClusterInput.DBClusterIdentifier
+    - CreateDBClusterInput.DomainIAMRoleName
+    - ModifyDBClusterInput.DomainIAMRoleName
+    - CreateDBClusterInput.KmsKeyId
```
```suggestion
```

You don't need to ignore referenced fields because we populate them with resolver result, hence it's fine for the code to use them. In custom struct, you need to add only the ref and selector fields.

hiteshghia-splunk replied:
So I noticed the auto-generated code output `KMSKeyID` but in the aws sdk structs its  `KmsKeyId`. I am guessing  conversion takes care of this?

hiteshghia replied:
So I noticed the auto-generated code output KMSKeyID but in the aws sdk structs its KmsKeyId. I am guessing conversion takes care of this?

muvaf reviewed (commented):
This is looking pretty good @hiteshghia ! I have a few comments but I think it's pretty close to merge. I have a few PRs up in code-generator that generates lateInitialize and isUpToDate (without this `Update` will never be called), so I can take care of those in another PR, no need to write them now.

hiteshghia-splunk commented:
> This is looking pretty good @hiteshghia ! I have a few comments but I think it's pretty close to merge. I have a few PRs up in code-generator that generates lateInitialize and isUpToDate (without this `Update` will never be called), so I can take care of those in another PR, no need to write them now.

@muvaf I was hoping to take a  stab at that as well next  after merging this, unless  there is rush to  get it in.

hiteshghia commented:
> This is looking pretty good @hiteshghia ! I have a few comments but I think it's pretty close to merge. I have a few PRs up in code-generator that generates lateInitialize and isUpToDate (without this `Update` will never be called), so I can take care of those in another PR, no need to write them now.

@muvaf I was hoping to take a stab at that as well next after merging this, unless there is rush to get it in, lmk.

hiteshghia commented:
@muvaf I have attempted to address your comments but facing multiple issue, I am going to attempt to frame the story of things that are going wrong :
- After the first commit in this PR, I could create a db cluster but the `READY` never updated to true for the managed resource and condition and the state/reason never updated to available. It was stuck in "creating" state. But in the aws console it was created/available fine.
- During ^^ time the provider would go into crashloop backoff because of NPE in zz_controller.go as it expects `filterList` to be init.
- So I went ahead added that in subsequent commit. Once I did that, the npe `filterList` went away but the provide reconciler loop would keep going into the `Create` method of the controller and keep erroring with DBCluster already exists.
- At this point I did a `make clean`, deleted the `.work` vendored folder and refreshed all the dependencies. The DBCluster already error kept happening.
- After which I decided to first fix the comments in the PR before starting to dig into the above error and stumbled into multiple different issues. I addressed all the comments in the PR in subsequent commits.
- I addressed https://github.com/crossplane/provider-aws/pull/546#discussion_r576658849 and https://github.com/crossplane/provider-aws/pull/546#discussion_r576659384 in that order.
- For the first pass for the VPCSecurityGroupIDs, I did not ignore it in the generator-config.yaml. But while creating the rds db cluster I started getting the following error:
`2021-02-28T17:25:47.029-0800	DEBUG	provider-aws	Cannot resolve managed resource references{"controller": "managed/dbcluster.rds.aws.crossplane.io", "request": "/example-rds", "uid": "8863eb4b-cace-4b36-bdbe-ee00a02df8b3", "version": "120207", "external-name": "example-rds", "error": "cannot update managed resource: DBCluster.rds.aws.crossplane.io \"example-rds\" is invalid: spec.forProvider.vpcSecurityGroupIDs: Invalid value: \"null\": spec.forProvider.vpcSecurityGroupIDs in body must be of type string: \"null\"", "errorVerbose": "DBCluster.rds.aws.crossplane.io \"example-rds\" is invalid: spec.forProvider.vpcSecurityGroupIDs: Invalid value: \"null\": spec.forProvider.vpcSecurityGroupIDs in body must be of type string: \"null\"\ncannot update managed resource\ngithub.com/crossplane/crossplane-runtime/pkg/reconciler/managed.(*APISimpleReferenceResolver).ResolveReferences\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/github.com/crossplane/crossplane-runtime@v0.12.1-0.20210219155338-30a941c3c3c6/pkg/reconciler/managed/api.go:147\ngithub.com/crossplane/crossplane-runtime/pkg/reconciler/managed.(*Reconciler).Reconcile\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/github.com/crossplane/crossplane-runtime@v0.12.1-0.20210219155338-30a941c3c3c6/pkg/reconciler/managed/reconciler.go:551\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:293\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:248\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:99\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1373"}
2021-02-28T17:25:47.029-0800	DEBUG	controller-runtime.manager.events	Warning	{"object": {"kind":"DBCluster","name":"example-rds","uid":"8863eb4b-cace-4b36-bdbe-ee00a02df8b3","apiVersion":"rds.aws.crossplane.io/v1alpha1","resourceVersion":"120207"}, "reason": "CannotResolveResourceReferences", "message": "cannot update managed resource: DBCluster.rds.aws.crossplane.io \"example-rds\" is invalid: spec.forProvider.vpcSecurityGroupIDs: Invalid value: \"null\": spec.forProvider.vpcSecurityGroupIDs in body must be of type string: \"null\""}`
   

- At which point I added ignored   `CreateDBClusterInput.VpcSecurityGroupIds` and  `ModifyDBClusterInput.VpcSecurityGroupIds` in the generator-config and added it as a custom param. 
- After doing that, the above error about vpcSecurityGroupId being null went away but I started getting - 
`2021-02-28T17:58:22.289-0800	DEBUG	provider-aws	Cannot create external resource	{"controller": "managed/dbcluster.rds.aws.crossplane.io", "request": "/example-rds", "uid": "8863eb4b-cace-4b36-bdbe-ee00a02df8b3", "version": "123049", "external-name": "example-rds", "error": "cannot create DBCluster in AWS: InvalidParameterValue: The parameter MasterUserPassword must be provided and must not be blank.", "errorVerbose": "InvalidParameterValue: The parameter MasterUserPassword must be provided and must not be blank.\ncannot create DBCluster in AWS\ngithub.com/crossplane/provider-aws/pkg/clients.Wrap\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/pkg/clients/aws.go:644\ngithub.com/crossplane/provider-aws/pkg/controller/rds/dbcluster.(*external).Create\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/pkg/controller/rds/dbcluster/zz_controller.go:119\ngithub.com/crossplane/crossplane-runtime/pkg/reconciler/managed.(*Reconciler).Reconcile\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/github.com/crossplane/crossplane-runtime@v0.12.1-0.20210219155338-30a941c3c3c6/pkg/reconciler/managed/reconciler.go:670\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:293\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:248\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/Users/hghia/go/src/cd.splunkdev.com/crossplane/provider-aws/.work/pkg/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:99\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1373"}
2021-02-28T17:58:22.289-0800	DEBUG	controller-runtime.manager.events	Warning	{"object": {"kind":"DBCluster","name":"example-rds","uid":"8863eb4b-cace-4b36-bdbe-ee00a02df8b3","apiVersion":"rds.aws.crossplane.io/v1alpha1","resourceVersion":"123049"}, "reason": "CannotCreateExternalResource", "message": "cannot create DBCluster in AWS: InvalidParameterValue: The parameter MasterUserPassword must be provided and must not be blank."}`
- For adding referencer for the master user password all I did was ignored that field in generator config and added `MasterUserPasswordSecretRef *xpv1.SecretKeySelector `json:"masterUserPasswordSecretRef,omitempty"`` in custom_types.go. I did not resolve any reeferences in referencer.go, but I am not sure how to handle that. rdsInstance eg that you pointed to is not really using code gen, rather uses the internal rds client which has some logic in getPassword method...

Please advice on how to proceed. Thanks.
@muvaf.

muvaf commented on pkg/controller/rds/dbcluster/setup.go:
```diff
@@ -0,0 +1,93 @@
+	clusterIdentifier := aws.String(meta.GetExternalName(cr))
+	resp := svcsdk.DescribeDBClustersOutput{}
+	for _, dbCluster := range obj.DBClusters {
+		if dbCluster.DBClusterIdentifier == clusterIdentifier {
```
```suggestion
		if aws.StringValue(dbCluster.DBClusterIdentifier) == aws.StringValue(clusterIdentifier) {
```
Both of these are pointers, meaning equality check is done on their memory address. However, we should check if their values match. That's why it always says the resource is not found, hence try to create it over and over.

muvaf reviewed (commented):
I left a comment about the problem that causes to `Create` to be called over and over again, which also causes the resource to never be `READY`.

> For adding referencer for the master user password all I did was ignored that field in generator config and added MasterUserPasswordSecretRef *xpv1.SecretKeySelector json:"masterUserPasswordSecretRef,omitempty"`` in custom_types.go. I did not resolve any reeferences in referencer.go, but I am not sure how to handle that.

If you'd like to, you leave the referencers and secret ref as is and I can open another PR that implements their resolvers. I'll change the `GetPassword` implementation so that it can be used by both `DBCluster` and `RDSInstance`.

> Also I used 6dfca97 commit in code_generation repo as the latest master errors out for me.

The latest `main` branch solves the panic problem regarding `filterList` so I'd suggest regenerating using the latest version of code-generator.

Additionally, we need to add `FinalDBSnapshotIdentifier` and `SkipFinalSnapshot` fields to `CustomDBClusterParameters` and then use their values in `preDelete` so that deletion works.

muvaf commented on examples/rds/dbcluster.yaml:
```diff
@@ -0,0 +1,21 @@
+    name: hghia-psqlserver-crossplane2
+    namespace: crossplane-system
+
+#  aws rds create-db-cluster --db-cluster-identifier sample-cluster --engine aurora-postgresql \
+#  --master-username user-name --master-user-password password \
+#  --db-subnet-group-name mysubnetgroup --vpc-security-group-ids sg-c7e5b0d2
```
```suggestion
```

muvaf commented on examples/rds/rdsinstance.yaml:
```diff
@@ -0,0 +1,33 @@
+# wip, repurpose this attach instances to the dbcluster.
```
Yeah, that would be a good idea.

hiteshghia replied:
I could not get the rdsinstance eg working so I have removed it from here for now. Will continue to try and get it working but we dont want to block this from being merged.
fyi I am using this to test locally  
```
apiVersion: database.aws.crossplane.io/v1beta1
kind: RDSInstance
metadata:
  name: example-rds
spec:
  forProvider:
    allocatedStorage: 20
    autoMinorVersionUpgrade: false
    backupRetentionPeriod: 0
    caCertificateIdentifier: rds-ca-2019
    copyTagsToSnapshot: false
    dbInstanceClass: db.m1.small
    enableIAMDatabaseAuthentication: false
    enablePerformanceInsights: false
    engine: aurora-postgresql
    licenseModel: general-public-license
    multiAZ: true
    port: 3306
    preferredBackupWindow: 06:15-06:45
    preferredMaintenanceWindow: sat:09:21-sat:09:51
    publiclyAccessible: false
    storageEncrypted: false
    storageType: gp2
    dbClusterIdentifier: example-rds
    region: us-east-1
  providerConfigRef:
    name: default
```

and this is the error I get -

```2021-03-02T21:47:02.159-0800	DEBUG	controller-runtime.manager.events	Warning	{"object": {"kind":"RDSInstance","name":"example-rds","uid":"7b31021f-e8f4-4155-9c8b-4dd2b909fe52","apiVersion":"database.aws.crossplane.io/v1beta1","resourceVersion":"375236"}, "reason": "CannotCreateExternalResource", "message": "cannot create RDS instance: InvalidParameterCombination: The requested DB Instance will be a member of a DB Cluster. Set master user password for the DB Cluster."}```

hiteshghia commented on CODE_GENERATION.md:
```diff
@@ -19,6 +19,7 @@ This guide shows how to get a new resource support up and running step by step.
 
 This is the very first step. You need to clone [code generator](https://github.com/aws-controllers-k8s/code-generator) from AWS and run the following
 command in the root directory:
+TODO : This fails with the latest code_generator. It broke somewhere between [5d5d1df, efe1ea9)
```
@muvaf I just tried again to run codegen using latest code-generator main but it just errors out for me. Also I cannot see the `Issues` tab in that repo to open one. FYI to move ahead for now - I am using commit # 5d5d1df in code-generator and it work fine.

hiteshghia commented:
@muvaf ok I think I have addressed all comments. Two things to note - https://github.com/crossplane/provider-aws/pull/546#discussion_r586129671 and  https://github.com/crossplane/provider-aws/pull/546#discussion_r586058170

muvaf commented on CODE_GENERATION.md:
```diff
@@ -201,10 +201,9 @@ func preDelete(_ context.Context, cr *svcapitypes.Stage, obj *svcsdk.DeleteStage
-
+If the external-name is decided by AWS after the creation, then you
+need to inject `postCreate` to set the crossplane resource external-name to 
+the unique identifier of the resource, for eg see [`apigatewayv2`](https://github.com/crossplane/provider-aws/blob/master/pkg/controller/apigatewayv2/api/setup.go#L77)
```
```suggestion
the unique identifier of the resource, for eg see [`apigatewayv2`](https://github.com/crossplane/provider-aws/blob/7f29cf3/pkg/controller/apigatewayv2/api/setup.go#L77)
```

We need to pin to the current commit because that line `master` might change in the future.

muvaf commented on apis/rds/v1alpha1/custom_types.go:
```diff
@@ -0,0 +1,103 @@
+
+	// A list of EC2 VPC security groups to associate with this DB cluster.
+	VPCSecurityGroupIDs []string `json:"vpcSecurityGroupIDs,omitempty"`
+	// VPCSecurityGroupIDs []string `json:"vpcSecurityGroupIds,omitempty"`
```
```suggestion
```

muvaf commented on apis/rds/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,93 @@
+		Extract:      reference.ExternalName(),
+	})
+	if err != nil {
+		return errors.Wrap(err, "spec.forProvider.DomainIAMRoleName")
```
```suggestion
		return errors.Wrap(err, "spec.forProvider.domainIAMRoleName")
```

muvaf commented on apis/rds/v1alpha1/referencers.go:
```diff
@@ -0,0 +1,93 @@
+		Extract:      reference.ExternalName(),
+	})
+	if err != nil {
+		return errors.Wrap(err, "spec.forProvider.KMSKeyID")
```
```suggestion
		return errors.Wrap(err, "spec.forProvider.kmsKeyID")
```

muvaf commented on examples/rds/dbcluster.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: rds.aws.crossplane.io/v1alpha1
+kind: DBCluster
+metadata:
+  name: example-rds
```
```suggestion
  name: example-dbcluster
```

muvaf reviewed (approved):
LGTM. I had a few nitpicks. Please squash and I'll merge this. After merging, I'll fix the password secret reference in another PR.

muvaf commented:
Thanks a lot @hiteshghia ! I'll follow up with a PR to fill the missing parts.
